File: Python/20_ValidParentheses.py
class Solution:
    def CacheIsEmpty(cache) -> bool:
        if(len(cache) == 0):
            return True
        return False
    
    def MatchIsFound(endOfCache, currentSymbol) -> bool:
        circleMatches = (endOfCache == "(" and currentSymbol == ")")
        curlyMatches = (endOfCache == "{" and currentSymbol == "}")
        squareMatches = (endOfCache == "[" and currentSymbol == "]")

        if(circleMatches or curlyMatches or squareMatches):
            return True
        return False

    def isValid(self, s: str) -> bool:
        cache = []
        
        for i,currentSymbol in enumerate(s):
            if(not Solution.CacheIsEmpty(cache)):
                endOfCache = cache[-1]

                if(Solution.MatchIsFound(endOfCache, currentSymbol)):
                    cache.pop()
                else:
                    cache.append(currentSymbol)
            else:
                cache.append(currentSymbol)
        if(Solution.CacheIsEmpty(cache)):
            return True
        return False

File: C++/20_ValidParentheses.cpp
class Solution {
public:
    bool matchIsFound(char prevLetter, char currentLetter) {
        bool curvedMatches = (prevLetter == '(' && currentLetter == ')');
        bool curlyMatches = (prevLetter == '{' && currentLetter == '}');
        bool squareMatches = (prevLetter == '[' && currentLetter == ']');

        if(curvedMatches || curlyMatches || squareMatches) {
            return true;
        }
        return false;
    }
    bool isValid(string s) {
        stack<char> entireStack;
        
        for (int i = 0; i < s.length(); i++) {
            char currentLetter = s[i];

            if (i > 0 && !entireStack.empty()) {
                char prevLetter = entireStack.top();

                if (matchIsFound(prevLetter, currentLetter)) {
                    entireStack.pop();
                }
                else {
                    entireStack.push(currentLetter);
                }
            }
            else {
                entireStack.push(currentLetter);
            }
        }
        if (entireStack.empty()) {
            return true;
        }
        return false;
    }
};
